designdocs/schemas_protocols_intents.md

```diff
@@ -0,0 +1,642 @@
+# 📘 Design Discussion: Schema Protocols and Intent-Based Delegation in Frequency
+
+## 0. **Work in Progress** <a id="section_0"></a>
+
+Note: This document is a work in progress; specific implementation details and code examples exhibited herein are for
+illustrative purposes only. Once the various questions and concerns surfaced by this "pre-design" document have be
+answered satisfactorily, the document will be updated and expanded to include specific details related to the proposed
+implementation.
+
+## 1. **Background and Motivation** <a id="section_1"></a>
+
+In the current implementation, schemas are registered with immutable numeric identifiers (`SchemaId`) and describe the
+layout and storage semantics (e.g., Avro/Parquet formats, on-chain/off-chain storage). These schema IDs are used as
+references by clients and runtime modules alike, particularly in the delegation system defined by the `msa` pallet.
+
+Delegations currently allow a user to authorize a provider (e.g., an app or service) to act on their behalf, but this
+authorization is tightly bound to a specific `SchemaId`. This model has proven limiting in several ways:
+
+- **Coupling between schema versions and delegations**
+- **Schemas represent data format, not purpose**
+- **Lack of human-readable context**
+- **Rigid governance workflows**
+- **No concept of publisher authority**
+
+These limitations have motivated a re-architecture of the schema and delegation systems to introduce the concepts of:
+
+- **Named schemas** with version tracking
+- **Structured publishing under protocols**
+- **Intent-based delegation**
+- **More expressive APIs and storage models**
+
+## 2. **Design Goals** <a id="section_2"></a>
+
+This section outlines the key objectives that guide the redesign of Frequency's schema and delegation architecture.
+
+- **Schema Immutability**
+- **Protocol-Based Versioning**
+- **Minimal Delegation Churn**
+- **Intent Separation**
+- **Scoped Delegations**
+- **Protocol Ownership and Governance**
+- **On-Chain Efficiency**
+
+## 3. Glossary of Terms <a id="section_3"></a>
+
+- **protocol** - the top level of the tripartite nomenclature _protocol.schema@version_ (ex: _dsnp_ is the protocol in
+  _dsnp.broadcast@v2_)
+- **schema** - the second level of _protocol.schema@version_ (ex: _dsnp.broadcast_ resolves to a specific schema).
+  Schemas should always be referred to by their fully-qualfied name (ie, _dsnp.broadcast_, not just _broadcast_)
+- **version** - the third level of _protocol.schema@version_, resolves to a specific _minor_ iteration of a schema
+- **intent** - an on-chain primitive representing an abstract action or operation
```

Comment:
Looking for some comments on naming.

I was initially in favor of: Namespace->Protocol->Schema; however, apparently "Protocol" for the 2nd level was rejected...?

@wilwade suggested Protocol->Schema->SchemaVersion. However, already in conversations it's clear that Schema vs. SchemaVersion is very confusing.

I think perhaps we should keep the third level as it currently is, `Schema`, to avoid renaming/cognitive shift. I've got some suggestions here for the other levels:

Level 1: Namespace, Domain, Package, Module, Scope
Level 2: Model, Template, Component

Happy to consider others...

Reply:
Hmm... what if we did:
1. Domain -> analagous to ENS Domain; will be an easy conversion if we ever implement/move to ENS
2. MetaSchema
3. Schema